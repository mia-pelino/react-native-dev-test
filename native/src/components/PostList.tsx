import React from 'react';
import {
  FlatList,
  Text,
  StyleSheet,
  TouchableOpacity,
  View,
} from 'react-native';
import { formatSummary, formatDate } from '../stringUtils';
import { Post } from '../types/Post';

type Props = {
  postList: Post[];
  isFiltered: boolean;
  pressAuthorHandler: (name: string) => void;
  pressPostHandler: (title: string, body: string) => void;
  displayAllPostsHandler: () => void;
};

const PostList: React.FC<Props> = ({
  postList,
  isFiltered,
  pressAuthorHandler,
  pressPostHandler,
  displayAllPostsHandler,
}) => {
  return (
    <View>
      {isFiltered ? (
        <TouchableOpacity onPress={displayAllPostsHandler}>
          <Text style={styles.instruction}>
            Press here to show the full list again.
          </Text>
        </TouchableOpacity>
      ) : (
        <Text style={styles.instruction}>
          Press an author's name to only show their posts.
        </Text>
      )}
      <FlatList
        style={styles.list}
        data={postList}
        renderItem={({ item }) => (
          <TouchableOpacity
            style={styles.posts}
            onPress={() => {
              pressPostHandler(item.title, item.body);
            }}
          >
            <Text>Title: {item.title}</Text>
            <Text>Summary: {formatSummary(item.body)}</Text>
            <TouchableOpacity
              onPress={() => pressAuthorHandler(item.author?.name)}
            >
              <Text style={styles.author}>Author: {item.author?.name}</Text>
            </TouchableOpacity>
            <Text>Date: {formatDate(item.publishedAt?.toString())}</Text>
          </TouchableOpacity>
        )}
      />
    </View>
  );
};

export default PostList;

const styles = StyleSheet.create({
  list: {
    marginTop: 20,
    width: '100%',
    alignSelf: 'center',
  },
  posts: {
    marginBottom: 15,
    marginHorizontal: 10,
    backgroundColor: '#84DCC6',
    paddingHorizontal: 12,
    paddingVertical: 8,
    borderRadius: 5,
  },
  author: {
    color: '#000',
    fontWeight: 'bold',
  },
  instruction: {
    color: '#FFA69E',
    fontWeight: 'bold',
    alignSelf: 'center',
    marginTop: 20,
  },
});
