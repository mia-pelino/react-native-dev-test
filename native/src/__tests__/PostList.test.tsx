import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import renderer, { create, act } from 'react-test-renderer';
import PostList from '../components/PostList';

describe('<PostList />', () => {
  const postData = [
    {
      title: 'tater tots',
      body: 'po \n\n## ta **toes**',
      author: {
        name: 'samwise',
        id: '1',
      },
      id: 'a',
      publishedAt: new Date('01-02-2020'),
    },
    {
      title: 'mashed potatoes',
      body: 'boil \n\n## em',
      author: {
        name: 'gamgee',
        id: '2',
      },
      id: 'b',
      publishedAt: new Date('01-01-2020'),
    },
  ];

  describe('snapshot', () => {
    it('lists all posts and displays message to tap author for filtering when list is not filtered', async () => {
      const tree = renderer
        .create(
          <PostList
            postList={postData}
            isFiltered={false}
            pressAuthorHandler={() => {}}
            pressPostHandler={() => {}}
            displayAllPostsHandler={() => {}}
          />
        )
        .toJSON();

      await act(async () => {
        expect(tree).toMatchSnapshot();
      });
    });

    it('only lists post by author and displays message to view all posts when list is filtered', async () => {
      const tree = renderer
        .create(
          <PostList
            postList={[postData[0]]}
            isFiltered={true}
            pressAuthorHandler={() => {}}
            pressPostHandler={() => {}}
            displayAllPostsHandler={() => {}}
          />
        )
        .toJSON();

      await act(async () => {
        expect(tree).toMatchSnapshot();
      });
    });

    it('lists a single post list when loaded result size is one', async () => {
      const tree = renderer
        .create(
          <PostList
            postList={[postData[0]]}
            isFiltered={false}
            pressAuthorHandler={() => {}}
            pressPostHandler={() => {}}
            displayAllPostsHandler={() => {}}
          />
        )
        .toJSON();

      await act(async () => {
        expect(tree).toMatchSnapshot();
      });
    });

    it('shows an empty post list when loaded results are empty', async () => {
      const tree = renderer
        .create(
          <PostList
            postList={[]}
            isFiltered={false}
            pressAuthorHandler={() => {}}
            pressPostHandler={() => {}}
            displayAllPostsHandler={() => {}}
          />
        )
        .toJSON();

      await act(async () => {
        expect(tree).toMatchSnapshot();
      });
    });
  });

  describe('interactivity', () => {
    let testIsFiltered: boolean = false;
    let mockPressAuthor = jest.fn(() => (testIsFiltered = true));
    let mockPressPost = jest.fn();
    let mockDisplayAllPosts = jest.fn(() => (testIsFiltered = false));

    it('calls pressAuthorHandler when author is selected', () => {
      const wrapper = render(
        <PostList
          postList={postData}
          isFiltered={testIsFiltered}
          pressAuthorHandler={mockPressAuthor}
          pressPostHandler={mockPressPost}
          displayAllPostsHandler={mockDisplayAllPosts}
        />
      );

      fireEvent.press(wrapper.getByText('Author: samwise'));

      expect(mockPressAuthor).toHaveBeenCalledTimes(1);
    });

    it('calls pressPostHandler when author is selected', () => {
      const wrapper = render(
        <PostList
          postList={postData}
          isFiltered={testIsFiltered}
          pressAuthorHandler={mockPressAuthor}
          pressPostHandler={mockPressPost}
          displayAllPostsHandler={mockDisplayAllPosts}
        />
      );

      fireEvent.press(wrapper.getByText('Title: tater tots'));

      expect(mockPressPost).toHaveBeenCalledTimes(1);
    });

    it('calls displayAllPostsHandler when instruction is pressed', () => {
      const wrapper = render(
        <PostList
          postList={postData}
          isFiltered={testIsFiltered}
          pressAuthorHandler={mockPressAuthor}
          pressPostHandler={mockPressPost}
          displayAllPostsHandler={mockDisplayAllPosts}
        />
      );

      fireEvent.press(
        wrapper.getByText('Press here to show the full list again.')
      );

      expect(mockDisplayAllPosts).toHaveBeenCalled();
    });
  });
});
