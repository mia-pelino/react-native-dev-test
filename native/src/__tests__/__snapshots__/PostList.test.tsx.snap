// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostList /> snapshot lists a single post list when loaded result size is one 1`] = `
<View>
  <Text
    style={
      Object {
        "alignSelf": "center",
        "color": "#FFA69E",
        "fontWeight": "bold",
        "marginTop": 20,
      }
    }
  >
    Press an author's name to only show their posts.
  </Text>
  <RCTScrollView
    data={
      Array [
        Object {
          "author": Object {
            "id": "1",
            "name": "samwise",
          },
          "body": "po 

## ta **toes**",
          "id": "a",
          "publishedAt": 2020-01-02T08:00:00.000Z,
          "title": "tater tots",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "alignSelf": "center",
        "marginTop": 20,
        "width": "100%",
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#84DCC6",
              "borderRadius": 5,
              "marginBottom": 15,
              "marginHorizontal": 10,
              "opacity": 1,
              "paddingHorizontal": 12,
              "paddingVertical": 8,
            }
          }
        >
          <Text>
            Title: 
            tater tots
          </Text>
          <Text>
            Summary: 
            po ta toes
          </Text>
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#000",
                  "fontWeight": "bold",
                }
              }
            >
              Author: 
              samwise
            </Text>
          </View>
          <Text>
            Date: 
            January 2, 2020
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`<PostList /> snapshot lists all posts and displays message to tap author for filtering when list is not filtered 1`] = `
<View>
  <Text
    style={
      Object {
        "alignSelf": "center",
        "color": "#FFA69E",
        "fontWeight": "bold",
        "marginTop": 20,
      }
    }
  >
    Press an author's name to only show their posts.
  </Text>
  <RCTScrollView
    data={
      Array [
        Object {
          "author": Object {
            "id": "1",
            "name": "samwise",
          },
          "body": "po 

## ta **toes**",
          "id": "a",
          "publishedAt": 2020-01-02T08:00:00.000Z,
          "title": "tater tots",
        },
        Object {
          "author": Object {
            "id": "2",
            "name": "gamgee",
          },
          "body": "boil 

## em",
          "id": "b",
          "publishedAt": 2020-01-01T08:00:00.000Z,
          "title": "mashed potatoes",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "alignSelf": "center",
        "marginTop": 20,
        "width": "100%",
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#84DCC6",
              "borderRadius": 5,
              "marginBottom": 15,
              "marginHorizontal": 10,
              "opacity": 1,
              "paddingHorizontal": 12,
              "paddingVertical": 8,
            }
          }
        >
          <Text>
            Title: 
            tater tots
          </Text>
          <Text>
            Summary: 
            po ta toes
          </Text>
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#000",
                  "fontWeight": "bold",
                }
              }
            >
              Author: 
              samwise
            </Text>
          </View>
          <Text>
            Date: 
            January 2, 2020
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#84DCC6",
              "borderRadius": 5,
              "marginBottom": 15,
              "marginHorizontal": 10,
              "opacity": 1,
              "paddingHorizontal": 12,
              "paddingVertical": 8,
            }
          }
        >
          <Text>
            Title: 
            mashed potatoes
          </Text>
          <Text>
            Summary: 
            boil em
          </Text>
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#000",
                  "fontWeight": "bold",
                }
              }
            >
              Author: 
              gamgee
            </Text>
          </View>
          <Text>
            Date: 
            January 1, 2020
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`<PostList /> snapshot only lists post by author and displays message to view all posts when list is filtered 1`] = `
<View>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        Object {
          "alignSelf": "center",
          "color": "#FFA69E",
          "fontWeight": "bold",
          "marginTop": 20,
        }
      }
    >
      Press here to show the full list again.
    </Text>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "author": Object {
            "id": "1",
            "name": "samwise",
          },
          "body": "po 

## ta **toes**",
          "id": "a",
          "publishedAt": 2020-01-02T08:00:00.000Z,
          "title": "tater tots",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "alignSelf": "center",
        "marginTop": 20,
        "width": "100%",
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#84DCC6",
              "borderRadius": 5,
              "marginBottom": 15,
              "marginHorizontal": 10,
              "opacity": 1,
              "paddingHorizontal": 12,
              "paddingVertical": 8,
            }
          }
        >
          <Text>
            Title: 
            tater tots
          </Text>
          <Text>
            Summary: 
            po ta toes
          </Text>
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#000",
                  "fontWeight": "bold",
                }
              }
            >
              Author: 
              samwise
            </Text>
          </View>
          <Text>
            Date: 
            January 2, 2020
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`<PostList /> snapshot shows an empty post list when loaded results are empty 1`] = `
<View>
  <Text
    style={
      Object {
        "alignSelf": "center",
        "color": "#FFA69E",
        "fontWeight": "bold",
        "marginTop": 20,
      }
    }
  >
    Press an author's name to only show their posts.
  </Text>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "alignSelf": "center",
        "marginTop": 20,
        "width": "100%",
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</View>
`;
